File: tools/blob2c.py
import sys
import re
import os

TEMPLATE = """\
#include "../src/ivan.h"
#include "../src/ivan_blobs.h"
#include "v8.h"

using namespace v8;

namespace ivan {{
namespace blobs {{

{definitions}

Local<String> MainSource(Isolate* isolate) {{
  return ivan_value.ToStringChecked(isolate);
}};

void DefineJavaScript(Isolate* isolate, Local<Object> target) {{
  {initializers}
}}

}}
}}
"""

ONE_BYTE_STRING = """
static const uint8_t raw_{var}[] = {{ {data} }};
static struct : public v8::String::ExternalOneByteStringResource {{
  const char* data() const override {{
    return reinterpret_cast<const char*>(raw_{var});
  }}
  size_t length() const override {{ return arraysize(raw_{var}); }}
  void Dispose() override {{ /* Default calls `delete this`. */ }}
  v8::Local<v8::String> ToStringChecked(v8::Isolate* isolate) {{
    return v8::String::NewExternalOneByte(isolate, this).ToLocalChecked();
  }}
}} {var};
"""

TWO_BYTE_STRING = """
static const uint16_t raw_{var}[] = {{ {data} }};
static struct : public v8::String::ExternalStringResource {{
  const uint16_t* data() const override {{ return raw_{var}; }}
  size_t length() const override {{ return arraysize(raw_{var}); }}
  void Dispose() override {{ /* Default calls `delete this`. */ }}
  v8::Local<v8::String> ToStringChecked(v8::Isolate* isolate) {{
    return v8::String::NewExternalTwoByte(isolate, this).ToLocalChecked();
  }}
}} {var};
"""

UNSIGNED_CHAR_STRING = """
const unsigned char raw_{var}[] = {{ {data} }};
const char* {var} = reinterpret_cast<const char*>(raw_{var});
"""

INITIALIZER = """\
CHECK(target->Set(isolate->GetCurrentContext(),
    {key}.ToStringChecked(isolate),
    {value}.ToStringChecked(isolate)).FromJust());
"""


def ToCArray(elements, step=10):
  elements = list(elements)
  # slices = (elements[i:i + step] for i in range(0, len(elements), step))
  # slices = map(lambda s: ','.join(str(x) for x in s), slices)
  # return ',\n'.join(slices)
  return ','.join([str(x) for x in elements])


def ToCString(contents):
  return ToCArray(map(ord, contents), step=20)


def Render(var, data, raw=False):
  if raw:
    template = UNSIGNED_CHAR_STRING
    data = ','.join(map(hex, data))
  elif any(ord(c) > 127 for c in data):
    data = map(ord, data.decode('utf-8').encode('utf-16be'))
    data = [data[i] * 256 + data[i + 1] for i in xrange(0, len(data), 2)]
    data = ToCArray(data)
    template = TWO_BYTE_STRING
  else:
    template = ONE_BYTE_STRING
    data = ToCString(data)
  return template.format(var=var, data=data)


def Blob2C(sources, target):
  definitions = []
  initializers = []

  for name in sources:
    mode = 'rb' if name.endswith('.bin') else 'rt'
    with open(name, mode) as file:
      data = file.read()

      if '/' in name or '\\' in name:
        split = re.split('/|\\\\', name)
        name = '/'.join(split)

      raw = name.endswith('.bin')
      name = os.path.splitext(name)[0]
      name = re.sub(r'^\/?lib\/', '', name)
      name = re.sub(r'^\/?deps\/v8\/out.gn\/x64.release\/', '', name)
      var = name.replace('-', '_').replace('/', '_').replace('.', '_')
      key = '%s_key' % var
      value = '%s_value' % var

      if not raw:
        definitions.append(Render(key, name))
        initializers.append(INITIALIZER.format(key=key, value=value))
      definitions.append(Render(value, data, raw))

  with open(target, 'w') as output:
    output.write(TEMPLATE.format(definitions=''.join(definitions),
                                 initializers=''.join(initializers)))

if __name__ == '__main__':
  target = sys.argv[1]
  source_files = sys.argv[2:]
  Blob2C(source_files, target)

File: tools/blob2c.js
'use strict';

/* eslint-env node */

const { readFileSync, writeFileSync } = require('fs');
const path = require('path');

const [output, ...inputs] = process.argv.slice(2);

const TEMPLATE = `\
#include "../src/zero.h"
#include "../src/zero_blobs.h"
#include "v8.h"

using namespace v8;

namespace zero {
namespace blobs {

{definitions}

Local<String> MainSource(Isolate* isolate) {
  return zero_value.ToStringChecked(isolate);
};

void DefineJavaScript(Isolate* isolate, Local<Object> target) {
  {initializers}
}

}  // namespace blobs
}  // namespace zero
`;

const ONE_BYTE_STRING = `
static const uint8_t raw_{name}[] = { {data} };
static struct : public v8::String::ExternalOneByteStringResource {
  const char* data() const override {
    return reinterpret_cast<const char*>(raw_{name});
  }
  size_t length() const override { return arraysize(raw_{name}); }
  void Dispose() override { /* Default calls \`delete this\`. */ }
  v8::Local<v8::String> ToStringChecked(v8::Isolate* isolate) {
    return v8::String::NewExternalOneByte(isolate, this).ToLocalChecked();
  }
} {name};
`;

const TWO_BYTE_STRING = `
static const uint16_t raw_{name}[] = { {data} };
static struct : public v8::String::ExternalStringResource {
  const uint16_t* data() const override { return raw_{name}; }
  size_t length() const override { return arraysize(raw_{name}); }
  void Dispose() override { /* Default calls \`delete this\`. */ }
  v8::Local<v8::String> ToStringChecked(v8::Isolate* isolate) {
    return v8::String::NewExternalTwoByte(isolate, this).ToLocalChecked();
  }
} {name};
`;

const INITIALIZER = `\
CHECK(target->Set(isolate->GetCurrentContext(),
                  {key}.ToStringChecked(isolate),
                  {value}.ToStringChecked(isolate)).FromJust());
`;

const format = (template, args) => {
  for (const [name, value] of Object.entries(args)) {
    template = template.replace(new RegExp(`\\{${name}\\}`, 'g'), value);
  }
  return template;
};

const utf16be = (str) => {
  const buffer = Buffer.from(str, 'UCS-2');
  const l = buffer.length;
  if (l & 0x01) {
    throw new Error('uneven buffer length');
  }
  for (let i = 0; i < l; i += 2) {
    const e = buffer[i];
    buffer[i] = buffer[i + 1];
    buffer[i + 1] = e;
  }
  return buffer;
};

const render = (name, source) => {
  let template;
  let data = '';
  source = Buffer.from(source);
  if (source.some((c) => c > 127)) {
    template = TWO_BYTE_STRING;
    const utf16 = utf16be(source.toString());
    for (let i = 0; i < utf16.length; i += 2) {
      data += `${(utf16[i] * 256) + (utf16[i + 1] || 0)},`;
    }
  } else {
    template = ONE_BYTE_STRING;
    for (let i = 0; i < source.length; i += 1) {
      data += `${source[i]},`;
    }
  }

  return format(template, { name, data });
};

const definitions = [];
const initializers = [];

for (const input of inputs) {
  const source = readFileSync(path.resolve(process.cwd(), input), 'utf8');

  const name = input.split('.')[0].replace(/^lib\//, '');
  const varName = name.replace(/[^a-zA-Z0-9]/g, '_');
  const key = `${varName}_key`;
  const value = `${varName}_value`;

  definitions.push(render(key, name));
  definitions.push(render(value, source));
  initializers.push(format(INITIALIZER, { key, value }));
}

const final = format(TEMPLATE, {
  definitions: definitions.join(''),
  initializers: initializers.join(''),
});

writeFileSync(path.resolve(process.cwd(), output), final);
